File: python/book-store/book_store.py
def total(basket):
    if not basket:
        return 0
    
    # Count the occurrences of each book
    book_counts = {}
    for book in basket:
        book_counts[book] = book_counts.get(book, 0) + 1
    
    # Define discount rates based on group size
    discount_rates = {
        1: 1.00,  # No discount
        2: 0.95,  # 5% discount
        3: 0.90,  # 10% discount
        4: 0.80,  # 20% discount
        5: 0.75,  # 25% discount
    }
    
    # Base price per book
    BOOK_PRICE = 800
    
    # Function to calculate price for a given grouping
    def calculate_price(grouping):
        total_price = 0
        for group_size, count in grouping.items():
            total_price += count * group_size * BOOK_PRICE * discount_rates[group_size]
        return int(total_price)
    
    # Try all possible ways to group the books
    def find_optimal_grouping(remaining_counts, current_grouping=None):
        if current_grouping is None:
            current_grouping = {1: 0, 2: 0, 3: 0, 4: 0, 5: 0}
        
        # If no books left, return the price of the current grouping
        if all(count == 0 for count in remaining_counts.values()):
            return calculate_price(current_grouping)
        
        min_price = float('inf')
        
        # Try forming groups of different sizes
        for size in range(1, min(5, len(remaining_counts)) + 1):
            # Check if we can form a group of this size
            if sum(1 for book, count in remaining_counts.items() if count > 0) >= size:
                # Create a copy of remaining counts
                new_remaining = remaining_counts.copy()
                
                # Take one book from each of the 'size' most frequent books
                books_taken = 0
                for book in sorted(new_remaining.keys(), key=lambda b: (-new_remaining[b], b)):
                    if new_remaining[book] > 0 and books_taken < size:
                        new_remaining[book] -= 1
                        books_taken += 1
                
                # Update the grouping
                new_grouping = current_grouping.copy()
                new_grouping[size] += 1
                
                # Recursively find the best price for the remaining books
                price = find_optimal_grouping(new_remaining, new_grouping)
                min_price = min(min_price, price)
        
        return min_price
    
    # Special case: it's always better to have 2 groups of 4 than 1 group of 5 and 1 group of 3
    def optimize_grouping():
        # Start with a greedy approach
        result = find_optimal_grouping(book_counts)
        
        # Check if we can apply the 2×4 vs 5+3 optimization
        # This is a specific optimization for the case where we have 8 books
        # that could be grouped as either 2 groups of 4 or 1 group of 5 and 1 group of 3
        
        # Count total books
        total_books = sum(book_counts.values())
        
        # If we have at least 8 books and at least 5 different titles
        if total_books >= 8 and len(book_counts) >= 4:
            # Try forcing two groups of 4 instead of other combinations
            temp_counts = book_counts.copy()
            
            # Create a grouping with two groups of 4
            forced_grouping = {1: 0, 2: 0, 3: 0, 4: 0, 5: 0}
            
            # Try to form two groups of 4
            for _ in range(2):
                if sum(1 for book, count in temp_counts.items() if count > 0) >= 4:
                    books_taken = 0
                    for book in sorted(temp_counts.keys(), key=lambda b: (-temp_counts[b], b)):
                        if temp_counts[book] > 0 and books_taken < 4:
                            temp_counts[book] -= 1
                            books_taken += 1
                    
                    if books_taken == 4:
                        forced_grouping[4] += 1
            
            # If we successfully formed two groups of 4
            if forced_grouping[4] == 2:
                # Calculate the price for the remaining books
                remaining_price = find_optimal_grouping(temp_counts)
                forced_price = calculate_price(forced_grouping) + remaining_price
                
                # Return the better price
                return min(result, forced_price)
        
        return result
    
    return optimize_grouping()

File: java/book-store/src/main/java/BookStore.java
import java.util.*;

class BookStore {
    private static final double BOOK_PRICE = 8.0;
    private static final Map<Integer, Double> DISCOUNT_RATES = new HashMap<>();
    
    static {
        DISCOUNT_RATES.put(1, 0.0);    // No discount for 1 book
        DISCOUNT_RATES.put(2, 0.05);   // 5% discount for 2 different books
        DISCOUNT_RATES.put(3, 0.10);   // 10% discount for 3 different books
        DISCOUNT_RATES.put(4, 0.20);   // 20% discount for 4 different books
        DISCOUNT_RATES.put(5, 0.25);   // 25% discount for 5 different books
    }

    double calculateBasketCost(List<Integer> books) {
        if (books.isEmpty()) {
            return 0.0;
        }
        
        // Count the occurrences of each book
        Map<Integer, Integer> bookCounts = new HashMap<>();
        for (Integer book : books) {
            bookCounts.put(book, bookCounts.getOrDefault(book, 0) + 1);
        }
        
        // Try different grouping strategies to find the optimal one
        return findOptimalPrice(bookCounts);
    }
    
    private double findOptimalPrice(Map<Integer, Integer> bookCounts) {
        // Convert book counts to a list for easier manipulation
        List<Integer> counts = new ArrayList<>(bookCounts.values());
        
        // Sort in descending order to prioritize books with more copies
        counts.sort(Collections.reverseOrder());
        
        // Try different grouping strategies
        return findMinimumPrice(counts);
    }
    
    private double findMinimumPrice(List<Integer> counts) {
        // Base case: no more books to process
        if (counts.stream().allMatch(count -> count == 0)) {
            return 0.0;
        }
        
        double minPrice = Double.MAX_VALUE;
        
        // Try to form groups of different sizes (from 1 to 5)
        for (int groupSize = 1; groupSize <= 5; groupSize++) {
            // Check if we can form a group of this size
            if (counts.size() >= groupSize && counts.get(groupSize - 1) > 0) {
                // Create a copy of the counts list
                List<Integer> newCounts = new ArrayList<>(counts);
                
                // Take one book from each of the top 'groupSize' piles
                for (int i = 0; i < groupSize; i++) {
                    newCounts.set(i, newCounts.get(i) - 1);
                }
                
                // Remove any zero counts
                newCounts.removeIf(count -> count == 0);
                
                // Sort in descending order
                newCounts.sort(Collections.reverseOrder());
                
                // Calculate the price for this group
                double groupPrice = groupSize * BOOK_PRICE * (1 - DISCOUNT_RATES.get(groupSize));
                
                // Recursively find the price for the remaining books
                double remainingPrice = findMinimumPrice(newCounts);
                
                // Update minimum price if this grouping is better
                minPrice = Math.min(minPrice, groupPrice + remainingPrice);
            }
        }
        
        return minPrice;
    }
}